lang_scala/parsing/Lexer_scala.mll

```diff
@@ -87,9 +87,16 @@ let pop_mode () = ignore(Common2.pop2 _mode_stack)
 
 let whiteSpace = [' ' '\t'] (* todo: also OA, OD? *)
 
-let upper = ['A'-'Z''$'] (* todo: and unicode category *)
-let lower = ['a'-'z''_'] (* todo: and unicode category L1 *)
-let letter = upper | lower (* todo: and unicode category Lo, Lt, Nl *)
+
+(* To allow for unicode characters, we just enumerate everything that a upper/lower is not.
+   This is actually more permissive than scala, which limits identifiers to certain classes of
+   unicode characters, but this is fine.
+*)
+let upper =
+  [^ 'a'-'z' '0'-'9' '+' '-' '*' '%' ':' ';' '=' '!' '#' '~' '?' '\\' '@' '|' '&' '^' '<' '>' ' ' '\t' '\n' '\r' '(' ')' '[' ']' '{' '}' '.' '/' ',' '"' '\'' '`']
+let lower =
+  [^ 'A'-'Z' '0'-'9' '+' '-' '*' '%' ':' ';' '=' '!' '#' '~' '?' '\\' '@' '|' '&' '^' '<' '>' ' ' '\t' '\n' '\r' '(' ')' '[' ']' '{' '}' '.' '/' ',' '"' '\'' '`']
+let letter = upper | lower
```

Comment:
You can use `#` for set difference on sets of characters defined with `[...]` or `_` or character literals. For example, the following example is valid:
```
let punct = [';' '!']
let lower = ['a'-'z']
let other = _ # punct # lower

rule main = parse
other as c { Char.code c }
```

Reply:
Ah this is good to know, I switched to this.